File: ImplusPurchasesSDK/ViewController.m
//
//  ViewController.m
//  ImplusPurchasesSDK
//
//  Created by Vilius Padleckis on 2020-09-16.
//  Copyright © 2020 Implus.co. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@end

@implementation ViewController



- (void)viewDidLoad {
    [super viewDidLoad];
    _tableView.delegate = self;
    _tableView.dataSource = self;
    [[IMPurchase shared] invalidateCache];
    [[IMPurchase shared] getProductGroups:^(NSDictionary<NSString *,IMGroup *> * _Nullable groups, NSError * _Nullable error) {
        self->_groups = groups;
        [self->_tableView reloadData];
    }];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self checkStatus];
    _activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    _activityIndicator.alpha = 1.0;
    _activityIndicator.center = self.view.center;
    [self.view addSubview:_activityIndicator];
    [_activityIndicator startAnimating];
    [_activityIndicator setHidden:true];
}

- (IBAction)buttonAction:(id)sender {
//    BOOL *sandbox = [[IMPurchase shared] isSandbox];
//    NSLog(@"%@",sandbox? @"YES":@"NO");
    [self checkStatus];
}
- (IBAction)invalidateCacheAction:(id)sender {
    [[IMPurchase shared] invalidateCache];
}

- (IBAction)restoreAction:(id)sender {
    [[IMPurchase shared] restorePurchases:^(NSDictionary<NSString *, IMEntitlement*>* ent, NSError * _Nullable error) {
        NSLog(@"restored");
    }];
}

-(void)checkStatus {
    [[IMPurchase shared] hasActiveSubscription:^(NSDictionary<NSString *,IMEntitlement *> * _Nullable entitlements, NSError * _Nullable error) {
        NSString * status = @"";
        if ([entitlements[@"pro"] isActive]) {
            status = @"pro active";
            [self->_statusLabel setTextColor:[UIColor greenColor]];
        } else {
            status = @"pro not active";
            [self->_statusLabel setTextColor:[UIColor redColor]];
        }
        [self->_statusLabel setText:[NSString stringWithFormat:@"Subcription status: %@", status]];
    }];
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSArray *allKeys = [_groups allKeys];
    NSString *curKey = [allKeys objectAtIndex:section];
    IMGroup *curArray = [_groups objectForKey:curKey];
    return [curArray.products count];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [_groups count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier: @"cell"];
    }
    NSArray *allKeys = [_groups allKeys];
    NSString *curKey = [allKeys objectAtIndex:indexPath.section];
    IMGroup *curArray = [_groups objectForKey:curKey];
    
    SKProduct* product = curArray.products[indexPath.row].product;
    NSString* indef = [product productIdentifier];
    
    NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
    [numberFormatter setFormatterBehavior:NSNumberFormatterBehavior10_4];
    [numberFormatter setNumberStyle:NSNumberFormatterCurrencyStyle];
    [numberFormatter setLocale:product.priceLocale];
    NSString *priceString = [numberFormatter stringFromNumber:product.price];
    NSString* isTrial = @"NO";
    if ([[product introductoryPrice] numberOfPeriods] != nil) {
        isTrial = @"Yes";
    }
    
    [[cell textLabel] setText:[NSString stringWithFormat:@"%@ | %@ | trial: %@", indef, priceString, isTrial]];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSArray *allKeys = [_groups allKeys];
    NSString *curKey = [allKeys objectAtIndex:indexPath.section];
    IMGroup *curArray = [_groups objectForKey:curKey];

    [_activityIndicator setHidden:false];
    [[IMPurchase shared] purchaseProduct:curArray.products[indexPath.row] completion:^(SKPaymentTransaction * _Nonnull transaction, NSDictionary<NSString *, IMEntitlement*>* _Nullable ent, NSError * _Nullable error, BOOL canceled) {
        if (canceled || error != nil) {
            NSLog(@"Canceled");
            [self->_errorLabel setText:error.localizedDescription];
//            if (error != nil) {
//                UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Error then purching"
//                                                                               message:error.description
//                                               preferredStyle:UIAlertControllerStyleAlert];
//
//                UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
//                   handler:^(UIAlertAction * action) {}];
//
//                [alert addAction:defaultAction];
//                [self presentViewController:alert animated:YES completion:nil];
//            }
        } else {
            NSLog(@"Purchase completed");
            [self checkStatus];
        }
        [_activityIndicator setHidden:true];
        
    }];
    
    [[tableView cellForRowAtIndexPath:indexPath] setSelected:false];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSArray *allKeys = [_groups allKeys];
    return [allKeys objectAtIndex:section];
}


@end

File: ImplusPurchasesExample/ViewController.swift
//
//  ViewController.swift
//  ImplusPurchasesExample
//
//  Created by Vilius Padleckis on 2020-09-28.
//  Copyright © 2020 Implus.co. All rights reserved.
//

import UIKit
import ImplusPurchases
class ViewController: UIViewController {

    @IBOutlet var tableView: UITableView!
    @IBOutlet var statusLabel: UILabel!
    @IBOutlet var errorLabel: UILabel!
    var activityIndicator: UIActivityIndicatorView!
    var groups: ProductGroups! = ProductGroups()
    let entitlementName = "pro"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.delegate = self
        tableView.dataSource = self
        IMPurchase.shared.invalidateCache()
        IMPurchase.shared.getProductGroups { (groups, error) in
            self.groups = groups
            self.tableView.reloadData()
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        checkStatus()
        activityIndicator = UIActivityIndicatorView(style: .large)
        activityIndicator.alpha = 1.0
        activityIndicator.center = self.view.center
        self.view.addSubview(activityIndicator)
        activityIndicator.startAnimating()
        activityIndicator.isHidden = true
    }
    
    @IBAction func checkSubAction(_ sender: Any) {
        checkStatus()
    }
    
    @IBAction func invalidateCacheAction(_ sender: Any) {
        IMPurchase.shared.invalidateCache()
    }
    
    @IBAction func restoreAction(_ sender: Any) {
        IMPurchase.shared.restorePurchases { (entitlement, error) in
            NSLog("restored");
            if let error = error {
                self.errorLabel.text = error.localizedDescription
            }
            if entitlement[self.entitlementName]?.isActive() == true {
                self.statusLabel.text = "Restore completed. Active purchase found"
            } else {
                self.statusLabel.text = "Restore completed. Active purchase not found"
            }
            self.statusLabel.textColor = UIColor.blue
        }
    }
    
    func checkStatus() {
        IMPurchase.shared.hasActiveSubscription { (entitlements, error) in
            var status = ""
            if entitlements[self.entitlementName]?.isActive() == true {
                status = "\(self.entitlementName) active"
                self.statusLabel.textColor = UIColor.green
            } else {
                status = "\(self.entitlementName) not active"
                self.statusLabel.textColor = UIColor.red
            }
            self.statusLabel.text = "Subscription status: \(status)"
        }
    }
    
    func purchaseProduct(product: IMProduct) {
        activityIndicator.isHidden = false
        IMPurchase.shared.purchaseProduct(product) { (_, entitlement, error, isCanceled) in
            if error != nil {
                if !isCanceled {
                    self.errorLabel.text = error!.localizedDescription
                }
            } else {
                print("Purchases completed")
                self.checkStatus()
            }
            self.activityIndicator.isHidden = true
        }
    }
}

extension ViewController: UITableViewDelegate, UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return groups.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let keys = Array(groups.keys)
        return groups[keys[section]]?.products.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let keys = Array(groups.keys)
        return keys[section]
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier: "cell")
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: "cell")
        }
        let keys = Array(groups.keys)
        let group = groups[keys[indexPath.section]]
        if let products = group?.products {
            let skProduct = products[indexPath.row].product
            let identifier = skProduct.productIdentifier
            var isTrial = "NO"
            
            let numberFormatter = NumberFormatter()
            numberFormatter.formatterBehavior = .behavior10_4
            numberFormatter.numberStyle = .currency
            numberFormatter.locale = skProduct.priceLocale
            let priceString = numberFormatter.string(from: skProduct.price) ?? "-"
            
            if skProduct.introductoryPrice?.numberOfPeriods != nil {
                isTrial = "YES"
            }
            
            cell!.textLabel?.text = "\(identifier) | \(priceString) | trial: \(isTrial)"
        } else {
            cell!.textLabel?.text = "Missing product data"
        }
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let keys = Array(groups.keys)
        guard let group = groups[keys[indexPath.section]] else { return }
        let product = group.products[indexPath.row]
        self.purchaseProduct(product: product)
    }
    
}

